homeassistant/components/iskra/config_flow.py

```diff
@@ -0,0 +1,256 @@
+                try:
+                    device_info = await test_rest_api_connection(self.host, user_input)
+                except CannotConnect:
+                    errors["base"] = "cannot_connect"
+                except NotAuthorised:
+                    # Proceed to authentication step.
+                    return await self.async_step_authentication()
+                except UnknownException:
+                    errors["base"] = "unknown"
+                    # If the connection was not successful, show an error.
+
+                # If the connection was successful, create the device.
+                if not errors:
+                    return await self._create_entry(
+                        host=self.host,
+                        protocol=self.protocol,
+                        device_info=device_info,
+                        user_input=user_input,
+                    )
+
+            if self.protocol == "modbus_tcp":
+                # Proceed to modbus step.
+                return await self.async_step_modbus_tcp()
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=STEP_USER_DATA_SCHEMA,
+            errors=errors,
+        )
+
+    async def async_step_authentication(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the authentication step."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            try:
+                device_info = await test_rest_api_connection(self.host, user_input)
+            # If the connection failed, abort.
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
+            # If the authentication failed, show an error and authentication form again.
+            except NotAuthorised:
+                errors["base"] = "invalid_auth"
+            except UnknownException:
+                errors["base"] = "unknown"
+
+            # if the connection was successful, create the device.
+            if not errors:
+                return await self._create_entry(
+                    self.host,
+                    self.protocol,
+                    device_info=device_info,
+                    user_input=user_input,
+                )
+
+        # If there's no user_input or there was an error, show the authentication form again.
+        return self.async_show_form(
+            step_id="authentication",
+            data_schema=STEP_AUTHENTICATION_DATA_SCHEMA,
+            errors=errors,
+        )
+
+    async def async_step_modbus_tcp(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the Modbus TCP step."""
+        errors: dict[str, str] = {}
+
+        # If there's user_input, check the connection.
+        if user_input is not None:
+            if user_input[CONF_ADDRESS] not in range(1, 256):
+                errors["base"] = "address_range"
```

Comment:
You just added validation for the port, should we do the same for the address so we can elimate this check?

Reply:
i did try at first i used range in voluptuous scheme but what happened is HA changed it to slider instead of input field.
it works ok for larger numbers is there a way to set it to display input field?

Reply:
Euh yes there is, it's a NumberSelector, let me check

Reply:
```py
            vol.Optional(
                CONF_TEMPERATURE,
                description={"suggested_value": options.get(CONF_TEMPERATURE)},
                default=RECOMMENDED_TEMPERATURE,
            ): NumberSelector(NumberSelectorConfig(min=0, max=1, step=0.05)),
```

Reply:
okay this works, i changed it so it uses Number selector but it returns float so i had to add conversion to integer hope that's ok